routing/fseg_index.hpp

```diff
@@ -0,0 +1,129 @@
+#pragma once
+
+#include "routing/joint.hpp"
+
+#include "coding/reader.hpp"
+#include "coding/write_to_sink.hpp"
+
+#include "std/cstdint.hpp"
+#include "std/unordered_map.hpp"
+#include "std/vector.hpp"
+
+namespace routing
+{
+class RoadJointIds final
+{
+public:
+  JointId GetJointId(uint32_t segId) const
+  {
+    if (segId < m_jointIds.size())
+      return m_jointIds[segId];
+
+    return kInvalidJointId;
+  }
+
+  void AddJoint(uint32_t segId, JointId jointId)
+  {
+    if (segId >= m_jointIds.size())
+      m_jointIds.insert(m_jointIds.end(), segId + 1 - m_jointIds.size(), kInvalidJointId);
+
+    ASSERT_EQUAL(m_jointIds[segId], kInvalidJointId, ());
+    m_jointIds[segId] = jointId;
+  }
+
+  template <typename F>
+  void ForEachJoint(F && f) const
+  {
+    for (uint32_t segId = 0; segId < m_jointIds.size(); ++segId)
+    {
+      JointId const jointId = m_jointIds[segId];
+      if (jointId != kInvalidJointId)
+        f(segId, jointId);
+    }
+  }
+
+  size_t GetSize() const { return m_jointIds.size(); }
+
+  template <class TSink>
+  void Serialize(TSink & sink) const
+  {
+    WriteToSink(sink, static_cast<JointId>(m_jointIds.size()));
+    for (JointId jointId : m_jointIds)
+      WriteToSink(sink, jointId);
+  }
+
+  template <class TSource>
+  void Deserialize(TSource & src)
+  {
+    JointId const jointsSize = ReadPrimitiveFromSource<JointId>(src);
+    m_jointIds.reserve(jointsSize);
+    for (JointId i = 0; i < jointsSize; ++i)
+    {
+      JointId const jointId = ReadPrimitiveFromSource<JointId>(src);
+      m_jointIds.emplace_back(jointId);
+    }
+  }
+
+private:
+  vector<JointId> m_jointIds;
```

Comment:
Please add comment for this field. It's filled in a special way. Most of its items is equal to kInvalidJointId.